beacon/validator/src/test/java/tech/pegasys/teku/validator/coordinator/GraffitiBuilderTest.java

```diff
@@ -0,0 +1,424 @@
+      graffitiBuilder.onExecutionClientVersion(BESU_CLIENT_VERSION);
+      logCaptor.assertInfoLog(
+          "Default graffiti to use when building block without external VC: \"TK"
+              + TEKU_CLIENT_VERSION.commit().toUnprefixedHexString()
+              + "BUabcdef12\". "
+              + "To change check validator graffiti options.");
+    }
+  }
+
+  @Test
+  public void onExecutionClientVersion_shouldLogDefaultMergedGraffiti() {
+    this.graffitiBuilder =
+        new GraffitiBuilder(
+            clientGraffitiAppendFormat, Optional.of(Bytes32Parser.toBytes32(ASCII_GRAFFITI_20)));
+    try (final LogCaptor logCaptor = LogCaptor.forClass(EventLogger.class)) {
+      graffitiBuilder.onExecutionClientVersion(BESU_CLIENT_VERSION);
+      logCaptor.assertInfoLog(
+          "Default graffiti to use when building block without external VC: \"I've proposed ablock TK"
+              + TEKU_CLIENT_VERSION.commit().toUnprefixedHexString().substring(0, 2)
+              + "BUab\". "
+              + "To change check validator graffiti options.");
+    }
+  }
+
+  @Test
+  public void buildGraffiti_shouldNotFail() {
+    this.graffitiBuilder =
+        new GraffitiBuilder(clientGraffitiAppendFormat, userGraffiti) {
+          @Override
+          protected int calculateGraffitiLength(Optional<Bytes32> graffiti) {
+            throw new RuntimeException("");
+          }
+        };
+    assertThat(graffitiBuilder.buildGraffiti(Optional.empty())).isEqualTo(Bytes32.ZERO);
+    final Bytes32 graffiti = Bytes32Parser.toBytes32(ASCII_GRAFFITI_20);
+    assertThat(graffitiBuilder.buildGraffiti(Optional.of(graffiti))).isEqualTo(graffiti);
+  }
+
+  @ParameterizedTest()
+  @MethodSource("getBuildGraffitiFixtures")
+  public void buildGraffiti_shouldProvideCorrectOutput(
+      final ClientGraffitiAppendFormat clientGraffitiAppendFormat,
+      final Optional<String> maybeUserGraffiti,
+      final String expectedGraffiti) {
+    this.graffitiBuilder = new GraffitiBuilder(clientGraffitiAppendFormat, userGraffiti);
+    graffitiBuilder.onExecutionClientVersion(BESU_CLIENT_VERSION);
+    final Bytes32 expectedGraffitiBytes = Bytes32Parser.toBytes32(expectedGraffiti);
+    assertThat(
+            new String(
+                Arrays.copyOfRange(
+                    expectedGraffitiBytes.toArray(),
+                    0,
+                    32 - expectedGraffitiBytes.numberOfTrailingZeroBytes()),
+                StandardCharsets.UTF_8))
+        .isEqualTo(expectedGraffiti);
+    assertThat(graffitiBuilder.buildGraffiti(maybeUserGraffiti.map(Bytes32Parser::toBytes32)))
+        .isEqualTo(expectedGraffitiBytes);
+  }
+
+  @Test
+  public void extractGraffiti_shouldReturnCorrectGraffiti() {
+    assertThat(graffitiBuilder.extractGraffiti(Optional.empty(), 0)).isEqualTo("");
+    assertThat(
+            graffitiBuilder.extractGraffiti(
+                Optional.of(Bytes32Parser.toBytes32(asciiGraffiti0)), 0))
+        .isEqualTo("");
```

Comment:
same with most of this suite, we're putting a ton of tests into 1 test case, so breakage will mean that we only see first fail.